### PR TITLE
fix: refactor js code

### PR DESCRIPTION
**Pull request by kusyka911**
# some refactoring and optimization for JS code

## What problem does this PR solve?

1. Make single JS for theme initialization and switching. I think it's not so big and it is ok. This also prevent exposing data and functions to global scope.
2. Create empty `main.js` script what will be concatenated with other scripts.
3. Wrap `goToTop.js` to `load` event handler.
4. Optimize code of `scriptBodyEnd.html` to add scripts in more easy way.
5. ability to add custom scripts to the site via configuration

## Is this PR adding a new feature?
Add ability to add custom scripts to the site via configuration

## Is this PR related to any issue or discussion?
no

## PR Checklist

- [x] I have verified that the code works as described/as intended.
- [ ] This change adds a social icon which has a permissive license to use it.
- [x] This change **does not** include any external library/resources.
- [x] This change **does not** include any unrelated scripts (e.g. bash and python scripts).
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).

### CODE REVIEW

File: assets/js/theme.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        }
+    }
+
+    window.addEventListener('DOMContentLoaded', () => {
+        // Theme switch
+        const lamp = document.getElementById("mode");
+
+        lamp.addEventListener("click", () => toggleTheme());
+
+        // Blur the content when the menu is open
```
Comment (kusyka911) [inline assets/js/theme.js:67]: This can be replaced with CSS.
Unfortunately `:has()` is still experimental and disabled by default in `firefox` and unavailable in `firefox android`.
See:  [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/:has#browser_compatibility) and [caniuse](https://caniuse.com/?search=%3Ahas)

So I'm not sure is this a good idea to use it.



```scss
body:has(#menu-trigger:checked) {
    .wrapper {
      animation: 0.2s ease-in forwards blur;
      -webkit-animation: 0.2s ease-in forwards blur;
    }
}
```

File: assets/js/theme.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        const cbox = document.getElementById("menu-trigger");
+
+        cbox.addEventListener("change", function () {
+            const area = document.querySelector(".wrapper");
+            this.checked
+                ? area.classList.add("blurry")
+                : area.classList.remove("blurry");
+        });
```
Comment (kusyka911) [inline assets/js/theme.js:75]: This can be replaced with CSS.
Unfortunately `:has()` is still experimental and disabled by default in `firefox` and unavailable in `firefox android`.
See:  [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/:has#browser_compatibility) and [caniuse](https://caniuse.com/?search=%3Ahas)

So I'm not sure is this a good idea to use it.



```scss
// wrap with @media where #menu-trigger becomes visible
body:has(#menu-trigger:checked) {
    .wrapper {
      animation: 0.2s ease-in forwards blur;
      -webkit-animation: 0.2s ease-in forwards blur;
    }
}
```

**Comment by kusyka911** [comment]
Unfortunately page speed results are not very good with thees changes. I'll try to fix it. https://pagespeed.web.dev/analysis/https-deploy-preview-39--hugo-blog-awesome-netlify-app/6z48zc8ylu?form_factor=mobile

File: assets/js/theme.js (hunk 2 above)
Comment (hugo-sid) [reply assets/js/theme.js:75]: @kusyka911, first thanks for this PR.

Here is what I have to say regarding the use of CSS `:has()`:

image

Thanks for the caniuse link. Looks like the global support for this property is only 86%.
As a safety measure, I have set the following bars for myself:
- global support > 95% - good to go
- global support >= 90% - we can consider using this
- global support <90 - we should not consider using this

In conclusion, I feel we should not use this. It's not worth sacrificing browser support for the few lines of JS.

What do you say @kusyka911 ?

Comment (kusyka911) [reply assets/js/theme.js:75]: In general I agree, but please also do not forget about no-script environments :wink:
We can add polyfill but in this case it will also not work in no-script environments.
So I don't see any good reason to push this idea.

Comment (hugo-sid) [reply assets/js/theme.js:75]: > We can add polyfill but in this case it will also not work in no-script environments.

Nice observation 😄

File: layouts/partials/scriptsBodyEnd.html
Comment (hugo-sid) [inline layouts/partials/scriptsBodyEnd.html:1]: This is a great idea @kusyka911 .

I feel, bundling JS to reduce number of network requests is crucial for improving website performance, as each request adds latency and consumes bandwidth.

File: layouts/partials/scriptsBodyStart.html [hunk 3]
```diff
@@ -1,2 +1,7 @@
-{{ $script_main := resources.Get "js/themeLoader.js" | minify | fingerprint }}
-<script src="{{ $script_main.RelPermalink }}"></script>
+{{ if hugo.IsProduction }}
+{{ $theme_script := resources.Get "js/theme.js" | minify | fingerprint }}
+<script src="{{ $theme_script.RelPermalink }}" integrity="{{ $theme_script.Data.Integrity }}"></script>
+{{ else }}
+{{ $theme_script := resources.Get "js/theme.js" | fingerprint }}
```
Comment (hugo-sid) [inline layouts/partials/scriptsBodyStart.html:5]: @kusyka911, do you think `| fingerprint` is useful, even when we are not using the `integrity` attribute in the script tag?

**Review by hugo-sid — COMMENTED** [review]
@kusyka911 I have done a minor refactoring of the docs in `config.toml`.

I would love to hear your insight on the question relating to `| fingerprint `.

File: layouts/partials/scriptsBodyStart.html (hunk 3 above)
Comment (kusyka911) [reply layouts/partials/scriptsBodyStart.html:5]: Fingerprint will be added to filename. This will prevent JS caching by browser while you running website in development mode and making JS changes.

Comment (hugo-sid) [reply layouts/partials/scriptsBodyStart.html:5]: Thanks for your response @kusyka911 .

> This should prevent JS caching by browser 

I have another perspective though:

- If someone in concerned about caching, most likely they will disable by going it to Dev tools. This is because they are concerned not only about JS caching but everything else (CSS, other assets) as well. The computation involved in calculating hash (due to `| fingerprint`) is rendered useless in this case.
Considering this, the wastage of computation can be quite high In development environment since, file changes (and thus hash computation) can happen very frequently.
- If someone is **not** concerned about caching, the computation involved in calculating hash is anyways not going to be useful.

What do you have to say in this regard?

> P.S. I'm sorry for late response, right now I'm a little busy working on another small project.

No worries. That's aright.

Comment (hugo-sid) [reply layouts/partials/scriptsBodyStart.html]: I have proceeded with removing ` fingerprint` in development. However, please feel free to submit a PR, if you see any good use case.

**Review by hugo-sid — APPROVED** [review]
Thanks @kusyka911 for refactoring JS.

I appreciate your contribution.
